### DISCUSSION TITLE
TMC5160 UART OR Single wire support

### DISCUSSION THREAD

**Discussion by muhammadelmogy** [grblHAL/core#283] [Ideas]
I was wondering if the TMC5160 UART or single wire is supported or it is just the SPI mode?
According to the datasheet, the driver can be connected using single wire or 2 wire diff. signal with a "chained sequential addressing scheme".
this is mentioned in page 28 section 5.3
[TMC5160A Datasheet Rev1.14](https://www.trinamic.com/fileadmin/assets/Products/ICs_Documents/TMC5160A_Datasheet_Rev1.14.pdf)
I am looking forward to use this driver and I was hoping to use it in a UART mode with grblHAL!

**Comment by muhammadelmogy** [comment]
BTW, The driver in mind will be mounted directly on the motor back that's why SPI/I2C is not a good choice for this type of application and a differential signal for communication is the best way for doing this in my opinion!

**Comment by terjeio** [comment]
Only SPI mode is supported since the Silent StepSticks cannot be configured for UART mode without hacking.
However it should be relatively straightforward to modify the code to support UART mode - AFAIKT the datagram layout is the same as the TMC2209 one. Address configuration is different though - see section 5.4.

**Reply by muhammadelmogy** [reply]
If you think about it, The CNC motors drivers is always in a separate board in a noisy box full of wires and the StepSticks style driver are not really that popular option for driving steppers in CNC machines so in order to connect any TMC Drivers and configure them, you will have to connect them to the controller over wires which means using SPI is not really reliable.

Not to mention the case I mentioned earlier which the driver is mounted far away on the driver which will require long wires for communications .
I really hope I have the knowledge and time to modify the code and provide a 2 wire-UART Support but I never really done something like this before!

**Reply by terjeio** [reply]
So motors with rear mounted TMC5160 drivers are readily available? A quick ebay search only returned Stepsticks...
I can write the code if someone donate at least 2 (3 would be better) such motor/driver combos so I have a setup to verify the code against.

**Reply by muhammadelmogy** [reply]
I did not really meant that TMC5160 rear motor mount are readily available!! (although there are some on AliExpress or at least with in the same category! ) but without doubt a remote Stepper motor (not mounted on the same driver board like the StepStick) is readily available and  one option of them is simply the TMC5160!
I am building my own TMC5160 driver for driving stepper motors in a separate module which will have RS485 and Encoder inputs and support up to 48V.
I still in the design phase so it will take some time before I could have a working prototype!
If you are plaining on implementing this one day, it would be a good idea to support the encoder input and try to do something in the firmware with it (although I am not yet sure if the driver is capable of handling the encoder as I imagine or not)!
